src/ripple/app/main/GRPCServer.h

```diff
@@ -0,0 +1,241 @@
+
+    Application& app_;
+
+    std::string serverAddress_;
+
+    // typedef for function to bind a listener
+    // This is always of the form:
+    // rpc::v1::XRPLedgerAPIService::AsyncService::Request[RPC NAME]
+    template <class Request, class Response>
+    using BindListener = std::function<void(
+        rpc::v1::XRPLedgerAPIService::AsyncService&,
+        grpc::ServerContext*,
+        Request*,
+        grpc::ServerAsyncResponseWriter<Response>*,
+        grpc::CompletionQueue*,
+        grpc::ServerCompletionQueue*,
+        void*)>;
+
+    // typedef for actual handler (that populates a response)
+    // handlers are defined in rpc/GRPCHandlers.h
+    template <class Request, class Response>
+    using Handler = std::function<std::pair<Response, grpc::Status>(
+        RPC::GRPCContext<Request>&)>;
+
+public:
+    GRPCServerImpl(Application& app);
+
+    void
+    shutdown();
+
+    // setup the server and listeners
+    // returns true if server started successfully
+    bool
+    start();
+
+    // the main event loop
+    void
+    handleRpcs();
+
+    // Create a CallData object for each RPC. Return created objects in vector
+    std::vector<std::shared_ptr<Processor>>
+    setupListeners();
+
+private:
+    // Class encompasing the state and logic needed to serve a request.
+    template <class Request, class Response>
+    class CallData
+        : public Processor,
+          public std::enable_shared_from_this<CallData<Request, Response>>
+    {
+    private:
+        // The means of communication with the gRPC runtime for an asynchronous
+        // server.
+        rpc::v1::XRPLedgerAPIService::AsyncService& service_;
+
+        // The producer-consumer queue for asynchronous server notifications.
+        grpc::ServerCompletionQueue& cq_;
+
+        // Context for the rpc, allowing to tweak aspects of it such as the use
+        // of compression, authentication, as well as to send metadata back to
+        // the client.
+        grpc::ServerContext ctx_;
+
+        // true if finished processing request
+        bool finished_;
+
+        Application& app_;
+
+        // mutex for signaling abort
+        std::mutex mut_;
+
+        // whether the call should be aborted, due to server shutdown
+        bool aborted_;
+
+        // What we get from the client.
+        Request request_;
+
+        // What we send back to the client.
+        Response reply_;
+
+        // The means to get back to the client.
+        grpc::ServerAsyncResponseWriter<Response> responder_;
+
+        // Function that creates a listener for specific request type
+        BindListener<Request, Response> bindListener_;
+
+        // Function that processes a request
+        Handler<Request, Response> handler_;
+
+        // Condition required for this RPC
+        RPC::Condition requiredCondition_;
+
+        // Load type for this RPC
+        Resource::Charge loadType_;
+
+    public:
+        virtual ~CallData() = default;
```

Comment:
Recommend adding:
```
        CallData(const CallData&) = delete;
        CallData& operator=(const CallData&) = delete;
```
purely for readability purposes.  There is no change in functionality here.